File: mobile/modules/bluetooth-sdk/android/src/main/java/com/mentra/bluetoothsdk/ReleaseChangelog.kt
package com.mentra.bluetoothsdk

/** One coordinated Mentra release note authored in /changelogs/<version>.md. */
data class ReleaseChangelog(
    val version: String,
    val markdown: String,
)

internal object ReleaseChangelogCatalog {
    private val baseVersionPattern = Regex("^(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(?:[-+].*)?$")

    fun select(
        fromVersion: String?,
        toVersion: String?,
    ): List<ReleaseChangelog> {
        val fallbackTarget =
            GeneratedReleaseMetadata.FAMILY_BASE_VERSION.ifBlank {
                GENERATED_RELEASE_CHANGELOGS.firstOrNull()?.version.orEmpty()
            }
        if (fallbackTarget.isBlank() && toVersion == null) return emptyList()
        val target = baseVersion(toVersion ?: fallbackTarget, "toVersion")
        if (GENERATED_RELEASE_CHANGELOGS.none { it.version == target }) {
            throw BluetoothSdkException("missing_changelog", "No changelog is bundled for target version $target.")
        }
        if (fromVersion == null) return GENERATED_RELEASE_CHANGELOGS.filter { it.version == target }

        val source = baseVersion(fromVersion, "fromVersion")
        val direction = compareVersions(target, source)
        return GENERATED_RELEASE_CHANGELOGS.filter { entry ->
            when {
                direction == 0 -> entry.version == target
                direction > 0 -> compareVersions(entry.version, source) > 0 && compareVersions(entry.version, target) <= 0
                else -> compareVersions(entry.version, source) < 0 && compareVersions(entry.version, target) >= 0
            }
        }
    }

    private fun baseVersion(
        version: String,
        label: String,
    ): String {
        val match = baseVersionPattern.matchEntire(version.trim())
            ?: throw BluetoothSdkException(
                "invalid_changelog_version",
                "$label must be a semantic version such as 3.1.0, 3.1.0-dev.4, or 3.1.0-beta.2.",
            )
        return "${match.groupValues[1]}.${match.groupValues[2]}.${match.groupValues[3]}"
    }

    private fun compareVersions(
        left: String,
        right: String,
    ): Int {
        val a = left.split(".").map(String::toInt)
        val b = right.split(".").map(String::toInt)
        for (index in 0..2) {
            if (a[index] != b[index]) return a[index] - b[index]
        }
        return 0
    }
}

File: mobile/modules/bluetooth-sdk/ios/Source/ReleaseChangelog.swift
import Foundation

/// One coordinated Mentra release note authored in /changelogs/<version>.md.
public struct ReleaseChangelog: Equatable, Sendable {
    public let version: String
    public let markdown: String

    public init(version: String, markdown: String) {
        self.version = version
        self.markdown = markdown
    }
}

enum ReleaseChangelogCatalog {
    static func select(fromVersion: String?, toVersion: String?) throws -> [ReleaseChangelog] {
        let fallbackTarget = GeneratedReleaseMetadata.familyBaseVersion.isEmpty
            ? (generatedReleaseChangelogs.first?.version ?? "")
            : GeneratedReleaseMetadata.familyBaseVersion
        if fallbackTarget.isEmpty, toVersion == nil { return [] }
        let target = try baseVersion(toVersion ?? fallbackTarget, label: "toVersion")
        guard generatedReleaseChangelogs.contains(where: { $0.version == target }) else {
            throw BluetoothSdkError(code: "missing_changelog", message: "No changelog is bundled for target version \(target).")
        }
        guard let fromVersion else {
            return generatedReleaseChangelogs.filter { $0.version == target }
        }

        let source = try baseVersion(fromVersion, label: "fromVersion")
        let direction = compareVersions(target, source)
        return generatedReleaseChangelogs.filter { entry in
            if direction == 0 { return entry.version == target }
            if direction > 0 {
                return compareVersions(entry.version, source) > 0 && compareVersions(entry.version, target) <= 0
            }
            return compareVersions(entry.version, source) < 0 && compareVersions(entry.version, target) >= 0
        }
    }

    private static func baseVersion(_ version: String, label: String) throws -> String {
        let value = version.trimmingCharacters(in: .whitespacesAndNewlines)
        let suffixStart = value.firstIndex(where: { $0 == "-" || $0 == "+" }) ?? value.endIndex
        let components = value[..<suffixStart].split(separator: ".", omittingEmptySubsequences: false)
        let isValid = components.count == 3 && components.allSatisfy { component in
            !component.isEmpty
                && (component == "0" || !component.hasPrefix("0"))
                && component.utf8.allSatisfy { byte in byte >= 48 && byte <= 57 }
        }
        guard isValid else {
            throw BluetoothSdkError(
                code: "invalid_changelog_version",
                message: "\(label) must be a semantic version such as 3.1.0, 3.1.0-dev.4, or 3.1.0-beta.2."
            )
        }
        return components.joined(separator: ".")
    }

    private static func compareVersions(_ left: String, _ right: String) -> Int {
        let a = left.split(separator: ".").map { Int($0) ?? 0 }
        let b = right.split(separator: ".").map { Int($0) ?? 0 }
        for index in 0 ..< 3 where a[index] != b[index] {
            return a[index] - b[index]
        }
        return 0
    }
}
